### PR TITLE
資料夾分類 / Support collection or folder

### PR DESCRIPTION
**Pull request by 0140454**
Issue: #109 

To implement folder feature, we create `Folders` table and introduce `sequelize-hierarchy` module which will create `FoldersAncestors` to maintain relationship between folders.

Each user will have a root folder. If user is new registered or there is no root folder assigned to user, it will create it automatically.

Also, notes created before folder feature will show in root folder.

### CODE REVIEW

**Comment by david50407** [comment]
Server doesn't response after I register new user and login on this PR (commit fd89373), and the error log below:

```log
hackmdPostgres_1  | ERROR:  column "hierarchyLevel" of relation "Folders" does not exist at character 63
hackmdPostgres_1  | STATEMENT:  INSERT INTO "Folders" ("id","ownerId","createdAt","updatedAt","hierarchyLevel") VALUES ('698b6913-66e2-4c9a-95d4-652ba7ecdb92','d2528a5d-4158-49b2-a9b8-97e6b1fd8020','2017-10-17 03:04:16.572 +00:00','2017-10-17 03:04:16.572 +00:00',1) RETURNING *;
hackmd_1          | Unhandled rejection SequelizeDatabaseError: column "hierarchyLevel" of relation "Folders" does not exist
hackmd_1          |     at Query.formatError (/hackmd/node_modules/sequelize/lib/dialects/postgres/query.js:357:14)
hackmd_1          |     at Result.<anonymous> (/hackmd/node_modules/sequelize/lib/dialects/postgres/query.js:88:19)
hackmd_1          |     at emitOne (events.js:96:13)
hackmd_1          |     at Result.emit (events.js:188:7)
hackmd_1          |     at Result.Query.handleError (/hackmd/node_modules/pg/lib/query.js:163:8)
hackmd_1          |     at Client.<anonymous> (/hackmd/node_modules/pg/lib/client.js:188:26)
hackmd_1          |     at emitOne (events.js:96:13)
hackmd_1          |     at Connection.emit (events.js:188:7)
hackmd_1          |     at Socket.<anonymous> (/hackmd/node_modules/pg/lib/connection.js:133:12)
hackmd_1          |     at emitOne (events.js:96:13)
hackmd_1          |     at Socket.emit (events.js:188:7)
hackmd_1          |     at readableAddChunk (_stream_readable.js:176:18)
hackmd_1          |     at Socket.Readable.push (_stream_readable.js:134:10)
hackmd_1          |     at TCP.onread (net.js:547:20)
```

File: lib/migrations/20170205174620-support-folder.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    }).then(function () {
+      return queryInterface.createTable('Folders', {
+        id: {
+          type: Sequelize.UUID,
+          primaryKey: true
+        },
+        name: Sequelize.TEXT,
+        ownerId: Sequelize.UUID,
+        createdAt: Sequelize.DATE,
+        updatedAt: Sequelize.DATE
```
[the remark this answers is not in the document]
Comment (0140454) [reply lib/migrations/20170205174620-support-folder.js]: Yes, you're right. Thanks for your great investigation.
We will push new commit as soon as possible.

**Comment by SISheogorath** [comment]
Has someone tested this PR in detail? 

- How well does the migration work? 
- Are existing notes lost?
- Do functionalities, like slide mode or published notes, break?

**Comment by david50407** [comment]
When you remove a folder, all notes in that folder will gone instead of going to root folder.
And you cannot find them in folder view anymore...

**Comment by dada8397** [comment]
@david50407 Do you mean that move notes to root folder better than delete all notes in the folder?

**Comment by david50407** [comment]
Or maybe let user choose where these notes go while deleting folders.
(Another folder or delete them)
Hsin Ta Chan <notifications@github.com>於 2017年10月21日 週六，10:48寫道：

**Comment by dada8397** [comment]
@david50407
Thanks for your advice 😄

**Comment by SISheogorath** [comment]
I'm more asking for migrations from an existing instance and database.

I just checked the background of these changes and it seems very good. But of course we need to provide a proper migration, so we need to make sure that we can redirect and migrate the existing notes when we want to merge this feature

**Comment by 0140454** [comment]
@SISheogorath 
About migration from an existing database, we have tested it using the following steps.
1. Use ```master``` branch to build a clean system.
2. Login and create some notes.
3. Use sequelize to migrate database to ```feature/folder``` branch.
4. Check created notes whether be displayed in root folder.

And we build an example instance at https://intense-waters-94449.herokuapp.com/, everyone can test this feature through it.

**Comment by SISheogorath** [comment]
This is **AWESOME**!!! ❤️ 

I'll make sure this makes it into the upcoming release!

**Comment by syhsu** [comment]
To be honest I don't really need folders to arrange my docs, which is, in my opinion, quite contradicted to the usages of tags. But I do hope to have "view" options in the dashboard, like a "list" view or a "tag" view,  to have a kind-of bird view of docs I have, especially when hackmd adds docs you check to your dashboard automatically, which is a pretty handy feature too, BTW.

The current icon view is good, but when you have docs more than a certain number, browsing among icons is really a nightmare. A convenient and straight-forward solution is to provide an option to show a full tag list somewhere in the dashboard and a check box is in front of each tag. Then I could check up the tags I want to see, instead of searching in a tag search bar. Hope this doesn't offend the UI designers who made the tag search bar. At last, hackmd is awesome.

**Comment by dada8397** [comment]
Hi @syhsu 
Thanks for your advice 😄 
I wonder that are you talking about the currently "card view" and you want some "list view" to be more convenient? This can be another issue. I think we can make a button to switch whether we want "card view" or "list view".
About tag search has already been in the page, you can choose tags in "select tags..." area, and you can filter tags.

**Comment by SISheogorath** [comment]
Agreed, I also think that the directory view is more independent from the current history view.

Central tagging is also way more difficult in HackMD since tags are part of the document. But that's another story. For this PR I think we are on a good way and I would like to see it in 1.1.0-ce. But when there is a majority against this change we can also re-discuss it.

I think folders and directories allow to categorize notes by adding a main topic above. Tags are different by design since they allow to add one document into multiple categories.

**Comment by syhsu** [comment]
Hello @dada8397,
Thanks for your reply. A "list view" is very very very good enough for me.:)

And hello @SISheogorath,
Agreed here. The folder implementation is a direction worthy to do in Hackmd since the tags in Hackmd are part of the document. The folder implementation here can improve multiple document operations, such as deleting/copying/selecting multiple documents. 

As for document organising, I really think tags can totally replace folders and can provide better user experiences via searching. My point is that searching is more convenient and efficient than organizing since search engines have changed how people memorize things, transactive memory. Tags provide not only a "flat" view of documents but also a structure of scaffolds for tracking your documents/thoughts.

P.S. Maybe using what question I ask when I organize a doc is a better way to explain my idea. When organizing a document, I ask myself what tags I should make to allow me find it in the future, not where I should put it.

**Comment by SISheogorath** [comment]
To be honest, after all this time, and all the changes we made, I don't seeing this implemented this way any time soon. To be fair and not keeping anyone up with this, I'll close this PR.

You did an awesome work and it's really nice, my problem with it is just that the change is too invasive. Sorry for that. Hope you still enjoy working with use!

**Pull request closed without merging**